ios/NumberTileGame/F3HBridgeModule.m

```diff
@@ -0,0 +1,35 @@
+//
+//  F3HBridgeModule.m
+//  NumberTileGame
+//
+//  Created by Michael Prouty on 10/31/19.
+//
+
+#import "F3HBridgeModule.h"
+
+#import "HighScoreArchiver.h"
+
+static UIViewController *menuViewController;
+
+@implementation F3HBridgeModule
+
++ (UIViewController*)getMenuViewController {
+    return menuViewController;
+}
+
++ (void)setMenuViewController:(UIViewController*)viewController {
+    menuViewController = viewController;
+}
+
+RCT_EXPORT_MODULE();
+
+RCT_EXPORT_METHOD(exitHighScoreButtonTapped:(NSArray*)scores) {
+    NSLog(@"exitHighScoreButtonTapped called with scores %@!", scores);
+    [HighScoreArchiver writeScores:scores];
```

Comment:
Code smell: the name `exitHighScoreButtonTapped:` doesn't indicate what the parameter is. It should be named `exitHighScoreButtonTappedWithScoreList:`.

But then, why am I passing along the whole list of scores to a method that is dismissing the view? There should be a separation of concerns. Sometimes app architecture requires combining things like this, but I don't think that's the case here.

The way this is written, the high score list is written to disk _every time_ the high score list is viewed, even if it didn't change. And you've set it up in such a way that the high score view could make arbitrary changes to the list of scores, when in reality it can only do two things: clear the list, or don't change it at all.

This is poor design; you've shoehorned functionality into an existing place rather than think about what is required.

Reply:
> And you've set it up in such a way that the high score view could make arbitrary changes to the list of scores, when in reality it can only do two things: clear the list, or don't change it at all.

> This is poor design; you've shoehorned functionality into an existing place rather than think about what is required.

Actually, I was trying to project into the future a little bit and think about the requirements of a hypothetical piece of the actual app. For example, if we were rendering a logbook editor instead of a high score screen, I was imagining that at the end of the editor's life it would pass its complete edited state back to Objective-C to be written to core data.

If you remember, I brought this up specifically during FTL on Wednesday and we discussed that this thread will be talking to Realm directly, so it won't be necessary to pass a bunch of state back to core data.

Reply:
I need to look at the Realm API to get a better sense of how things might work. If RN can only pass JSON-serializable objects back and forth, then we’re probably going to use a pattern where each “side” (native and JS) has their own model objects loaded into memory and the sides are just passing signals back and forth like “hey I just updated object xyzzy, you should reload it”.

Reply:
Pasting Jimmy's comment onto appropriate thread:

If we need to go this route, I only ask that we don’t build this
notification system from scratch.

- Jimmy

Reply:
Putting the details of the communication aside, this should be two separate methods:

1. `updateScores:` or `clearHighScoreList` or something that is responsible for updating the data. (Model Layer)
2. `exitHighScoreButtonTapped` to dismiss the view controller. (Controller Layer)

There's no good reason for one method to do both jobs.